for-loops.md

```diff
@@ -0,0 +1,207 @@
+- Feature Name: for-loops
+- Start Date: 2021-01-20
+
+# Summary
+[summary]: #summary
+
+This RFC gives a concrete proposal for adding `for` loops to Dafny.
+
+# Motivation
+[motivation]: #motivation
+
+Dafny has only one kind of loop, `while` loops (with either one guard or with a set of guards). Some tasks
+where other languages require a loop can be done in Dafny with the aggregate `forall` statement, avoiding
+iteration altogether. Still, some situations call for a loop with a loop index and iteration pattern just
+like what a typical `for` loop provides. This RFC
+gives a concrete proposal, so that design aspects of `for` loops in the context of Dafny can be discussed.
+
+Two central design points are:     
+- the half-open interval of values for which the loop body is performed (e.g., `0 <= i < a.Length`) and
+  the closed interval that appears in the loop invariant (e.g., `0 <= i <= a.Length`)
+- what should be the possible increments of the loop index (e.g., `+1`, `-1`, `+2`)
+
+Requests for `for` loops in Dafny come up from time to time. Here is a record of
+[one such request](https://github.com/dafny-lang/dafny/issues/7).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `for` loop has the following syntax:
+
+```
+ForLoop ::=
+  "for" Ident [":" Type] ":=" Expression ("to" | "downto") Expression
+  { InvariantDecl | ModifiesDecl }
+  [ Body ]
+```
+
+The scope of the loop-index variable introduced by `Ident` is the `for` loop itself. Its name must not be
+the same as any local variable introduced in the same block scope as the `for` loop itself.
+The loop index is immutable in the `Body`.
+
+The type of loop index is `Type`, if present; otherwise, it is inferred from the types of the
+two expressions, which must be assignable to the loop index. In either case, the type must be an
+integer-based type (`int`, some subset type based on an integer-based type, or a `newtype` based
+on an integer-based type; note that bitvector types, `ORDINAL`, and `real`-based types are not
+allowed).
+
+The loop index is automatically updated by the `for` loop. This guarantees that the loop will always
+terminate. Therefore, the `for` loop does not allow user-specified `decreases` clauses.
+
+The meaning of the loop is defined in terms of existing language features. Ignoring the scope of the
+loop index (which was explained above), the "up" loop
+
+```
+for i: I := lo to hi
+  invariant J
+  modifies W
+{
+  Body
+}
+```
+
+has the meaning
+
+``` dafny
+var _lo := lo;
+var _hi := hi;
+assert _lo <= _hi;  // this is Dafny-like
```

Comment:
What does `// this is Dafny-like` mean?

Do we really need to assert `_lo <= _hi`? This could be needlessly restrictive: I think there are some situations where `lo` might already be bigger than `hi`, and the intended behavior is simply to do no loop iterations at all if that's the case.

Reply:
I definitely agree. It is really useful if the loop simply does nothing when `lo > hi`. Note: I understand it will do nothing when `lo == hi`, the interval being half-open.

Reply:
I thought this would be "Dafny-like" because Dafny already imposes the restriction `lo <= hi` in an expression `s[lo..hi]`.

Reply:
What is an example when allowing `lo > hi` is useful? Is it a frequent occurrence?

Reply:
I think `lo == hi` (which should result in 0 iterations) is very common, and `lo > hi` is less common, but here's an example: Say I write a method to resize an array from `oldsize` to `newsize` (and I don't know which one is bigger), and if `newsize` is bigger, the new elements should be initialized to some `default` value, so after copying the elements, maybe I want to be "smart" and write a loop like `for i in [oldsize..newsize] { newarray[i] = default }` and run that loop no matter whether `oldsize` or `newsize` is bigger, relying on the fact that if `oldsize` is bigger, 0 iterations will be performed. But I guess examples like this one can be easily rewritten so that they also work if `lo <= hi` is required by Dafny, I can't think of an example right now where such rewriting would be too cumbersome.

Reply:
Thanks for the thoughts. For this particular example, one can also use a `forall` statement:

``` dafny
forall i | 0 <= i < newsize {
  newarray[i] := if i < oldsize then oldarray[i] else default
}
```